### PR TITLE
[DPE-4763] COS basics

### PR DESCRIPTION
**Pull request by juditnovak**
Dashboards COS integration for:
 - Prometheus
 - Grafana
 - Loki
 - Alerting

Tested locally, currently only relation data correctness is tested here.

Prometheus:

Screenshot from 2024-07-12 19-44-33

Grafana:

Screenshot from 2024-07-19 01-02-23

Loki:

Screenshot from 2024-07-19 00-06-52

Alerts:
Screenshot from 2024-07-19 01-50-49

Screenshot from 2024-07-19 01-51-55

Screenshot from 2024-07-19 02-01-24

Screenshot from 2024-07-19 02-01-41

### CODE REVIEW

File: actions.yaml [hunk 1]
```diff
@@ -11,18 +11,6 @@ get-password:
 pre-upgrade-check:
   description: Run necessary pre-upgrade checks before executing a charm upgrade.
 
-set-password:
```
Comment (phvalguima) [inline actions.yaml:14]: No more set-password is possible?

Comment (phvalguima) [inline actions.yaml:14]: No more set-password is possible?

File: src/events/password_actions.py [hunk 2]
```diff
@@ -1,73 +0,0 @@
-#!/usr/bin/env python3
-# Copyright 2023 Canonical Ltd.
-# See LICENSE file for licensing details.
-
-"""Event handlers for password-related Juju Actions."""
```
Comment (phvalguima) [inline src/events/password_actions.py:5]: why going away?

File: tests/integration/test_charm.py [hunk 3]
```diff
@@ -209,27 +216,32 @@ async def test_dashboard_password_rotation(ops_test: OpsTest):
...
-    password = await get_user_password(ops_test, user)
-    assert len(password) == 32
-
-    leader_num = await get_leader_id(ops_test)
-
-    # Change both passwords
-    result = await set_password(ops_test, username=user, num_unit=leader_num)
-    assert f"{user}-password" in result.keys()
-
+async def test_cos_relations(ops_test: OpsTest):
```
Comment (phvalguima) [inline tests/integration/test_charm.py:219]: Missing the same test as we have on `opensearch`, where you `curl` directly the `/metrics` endpoints and see if there is data coming out.

**Review by phvalguima — COMMENTED** [review]
It is not clear for me why a COS PR is removing the `set-password`, or why password-related code is going away.

File: actions.yaml (hunk 1 above)
Comment (juditnovak) [reply actions.yaml:14]: No, we actually never could set a password (so far).

We can:
 - either extend Opensearch Dashboards to allow for setting Opensearch user passwords
 - or stick to the current approach were user passwords are to be set directly on Opensearch

(Note that Dashboards is only using Opensearch users. It doesn't have any users on its own.)

File: src/events/password_actions.py (hunk 2 above)
Comment (juditnovak) [reply src/events/password_actions.py:5]: This has never done anything else but setting a secret that the charm associated with a "fiktive" user that never existed.

File: tests/integration/test_charm.py (hunk 3 above)
Comment (juditnovak) [reply tests/integration/test_charm.py:219]: Note: this request introduces a dependency to the PR.

We can only implement that test after:
 - we have the snap deploying the exporter service
 - that snap has been released (note that we have major issues around dashboards snap releases)
 - that snap has been taken into the Dashboards code.

Fine with me, we can delay this code to that point. Alternatively we can have quick small PRs avoiding too much to pile up for last minute.

Comment (juditnovak) [reply tests/integration/test_charm.py:219]: Linked it against https://warthogs.atlassian.net/browse/DPE-4879 being 'blocked by', and moved the ticket to a BLOCKED state

File: actions.yaml (hunk 1 above)
Comment (phvalguima) [reply actions.yaml:14]: Yes, makes sense to remove it then

**Review by phvalguima — DISMISSED** [review]
Nice! LGTM

File: src/charm.py [hunk 4]
```diff
@@ -78,6 +76,17 @@ def __init__(self, *args):
...
 
+        # --- COS ---
+        self.cos_integration = COSAgentProvider(
+            self,
+            relation_name=COS_RELATION_NAME,
+            metrics_endpoints=[],
+            scrape_configs=self._scrape_config,
+            refresh_events=[self.on.config_changed],
+            # metrics_rules_dir="./src/alert_rules/prometheus",
+            # log_slots=["opensearch:logs"],
```
Comment (Mehdi-Bendriss) [inline src/charm.py:87]: With the patch you brough on the snap, this slot can be re set now I think `opensearch-dashboards:log`

File: src/charm.py [hunk 5]
```diff
@@ -218,11 +222,6 @@ def _restart(self, event: EventBase) -> None:
 
     def init_server(self):
         """Calls startup functions for server start."""
-        # don't run if leader has not yet created passwords
-        if not self.state.cluster.internal_user_credentials:
```
Comment (Mehdi-Bendriss) [inline src/charm.py:222]: Wouldn't this run and attempt to start the server even prior to having a relation with `opensearch` (and receiving kiabanserver)?

File: src/charm.py [hunk 6]
```diff
@@ -245,6 +244,19 @@ def init_server(self):
...
+    def _scrape_config(self) -> list[dict]:
+        """Generates the scrape config as needed."""
+        return [
+            {
+                "metrics_path": "/metrics",
+                "static_configs": [
+                    {"targets": [f"{self.state.unit_server.private_ip}:{COS_PORT}"]}
+                ],
+                # "tls_config": {"ca": self.state.unit_server.ca},
+                "scheme": "http",
```
Comment (Mehdi-Bendriss) [inline src/charm.py:256]: Should we add a conditional statement here? when TLS is enabled vs when it's not?

**Review by Mehdi-Bendriss — COMMENTED** [review]
Thanks Judit - it looks good. I only have a couple of questions to clarify

File: src/charm.py (hunk 4 above)
Comment (juditnovak) [reply src/charm.py:87]: Yes, I'm adding those changes all together once the snap version van be increased

File: src/charm.py (hunk 5 above)
Comment (juditnovak) [reply src/charm.py:222]: `kibanasever` was never returned by `self.state.cluster.internal_user_credentials`.

It was only `CHARM_USERS` that contained `monitor` only. (Note `monitor` user never existed, we just "pretended" it's there by setting a `monitor-password` for it. Has never been created for real.)

File: src/charm.py (hunk 6 above)
Comment (juditnovak) [reply src/charm.py:256]: `kibana-exporter` is only available HTTP :-/

Comment (juditnovak) [reply src/charm.py:256]: (NOTE: The whole COS service alltogether is only over HTTP, meaning Prometheus, Kibana, etc. interfaces. So I think w'er not that bad with this :-) Even thought indeed it's a limitation.)

File: src/charm.py (hunk 4 above)
Comment (juditnovak) [reply src/charm.py]: Done.

**Review by phvalguima — APPROVED** [review]

File: tests/integration/helpers.py [hunk 7]
```diff
@@ -139,6 +139,29 @@ def restart_unit(model_full_name: str, unit: str) -> None:
...
+    """Check if a given unit has 'kibana-exporter' service available and publishing."""
+    try:
+        # Normal IP address
+        socket.inet_aton(host)
+    except OSError:
+        socket.inet_pton(socket.AF_INET6, host)
+        host = f"[{host}]"
+
+    url = f"http://{host}:9684/metrics"
+    response = requests.get(url)
```
Comment (Mehdi-Bendriss) [inline tests/integration/helpers.py:152]: We need to handle http exceptions here - and return False if it fails

**Review by Mehdi-Bendriss — COMMENTED** [review]
Thanks Judit - great work! I have 1 comment

File: tests/integration/helpers.py (hunk 7 above)
Comment (juditnovak) [reply tests/integration/helpers.py]: Done.

**Review by Mehdi-Bendriss — APPROVED** [review]

**Comment by Mehdi-Bendriss** [comment]
The issues raised from the HA network cut tests will be solved in a future PR this pulse

**Pull request merged**